File: subprojects/gradle-core/src/main/groovy/org/gradle/util/Jvm.java
/*
 * Copyright 2010 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.gradle.util;

import org.apache.tools.ant.util.JavaEnvUtils;

import java.io.File;

public class Jvm {
    public static Jvm current() {
        return new Jvm();
    }

    Jvm() {
    }

    @Override
    public String toString() {
        return String.format("%s (%s %s)", System.getProperty("java.version"), System.getProperty("java.vm.vendor"), System.getProperty("java.vm.version"));
    }

    public File getJavaExecutable() {
        return new File(JavaEnvUtils.getJdkExecutable("java"));
    }

    public File getJavadocExecutable() {
        return new File(JavaEnvUtils.getJdkExecutable("javadoc"));
    }

    public boolean isJava5Compatible() {
        return System.getProperty("java.version").startsWith("1.5") || isJava6Compatible();
    }

    public boolean isJava6Compatible() {
        return System.getProperty("java.version").startsWith("1.6");
    }

    public File getToolsJar() {
        File javaHome = new File(System.getProperty("java.home"));
        File toolsJar = new File(javaHome, "/lib/tools.jar");
        if (toolsJar.exists()) {
            return toolsJar;
        }
        if (javaHome.getName().equalsIgnoreCase("jre")) {
            javaHome = javaHome.getParentFile();
            toolsJar = new File(javaHome + "/lib/tools.jar");
        }
        if (!toolsJar.exists()) {
            return null;
        }
        return toolsJar;
    }

}

File: .teamcity/src/main/kotlin/common/Jvm.kt
/*
 * Copyright 2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package common

interface Jvm {
    val version: JvmVersion
    val vendor: JvmVendor
}

data class DefaultJvm(
    override val version: JvmVersion,
    override val vendor: JvmVendor,
) : Jvm

object BuildToolBuildJvm : Jvm {
    override val version: JvmVersion
        get() = JvmVersion.JAVA_17
    override val vendor: JvmVendor
        get() = JvmVendor.OPENJDK
}

object OpenJdk8 : Jvm {
    override val version: JvmVersion
        get() = JvmVersion.JAVA_8
    override val vendor: JvmVendor
        get() = JvmVendor.OPENJDK
}

object OpenJdk11 : Jvm {
    override val version: JvmVersion
        get() = JvmVersion.JAVA_11
    override val vendor: JvmVendor
        get() = JvmVendor.OPENJDK
}

object OpenJdk17 : Jvm {
    override val version: JvmVersion
        get() = JvmVersion.JAVA_17
    override val vendor: JvmVendor
        get() = JvmVendor.OPENJDK
}
